File: SNN_VS_DEBUG/MIMO.c

#include "SNN.h"

int weighted_sum(double* weight, double* input, int LEN) {
	double output = 0;
	for (int i = 0; i < LEN; i++) {
		output += weight[i] * input[i];
	}
	return output;
}

double MIMO(double* input, double* LEN, double* output_vector, int LEN_O, int LEN_I, double weight[LEN_O][LEN_I]) {

	for (int i = 0; i < LEN_O; i++) {
		for (int j = 0; j < LEN_I; j++) {
			output_vector[j] += input[j]*weight[i][j];
		}
	}
}

File: algo/mimo/mimo.py

from wsgiref.util import request_uri


def mimo(input, weight):
    '''
    Args:
        input: input value in number.
        weight: input value in array of number.
    Returns:
            output return in float.
    Raises:
        ValueError: datatype error.
    '''

    if len(input) != len(weight):
        raise KeyError('input and weight must be the same length')
    
    if not isinstance(input, list) or not isinstance(weight, list):
        raise ValueError('input and weight must be list of number')    
    output = []

    for i in range(len(input)):
        if not isinstance(input[i], float):
            raise ValueError('input must be float')         
        output.append(elementwise_multiply(input, weight[i]))
    
    return output


def elementwise_multiply(input, weight):
    
    if len(input) != len(weight):
        raise KeyError('in elementwise_multiply input and weight must be the same length')

    output = 0
    for i in range(len(input)):
        if not isinstance(input[i], float) or not isinstance(weight[i], float):
            raise ValueError('in elementwise_multiply input and weight must be float')
        output += input[i] * weight[i]

    return output